pkgs/applications/networking/browsers/microsoft-edge/update.py

```diff
@@ -0,0 +1,52 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -i python3 -p python3Packages.packaging python3Packages.debian
+
+from urllib import request
+import gzip
+import base64
+
+from debian.deb822 import Packages
+from debian.debian_support import Version
+
+# Download and extract the package list file
+packages_url = 'https://packages.microsoft.com/repos/edge/dists/stable/main/binary-amd64/Packages.gz'
+handle = request.urlopen(packages_url)
+data = gzip.GzipFile(fileobj=handle).read()
+
+# Find latest version for each channel
+latest_packages: dict[str, Packages] = {}
+for package in Packages.iter_paragraphs(data, use_apt_pkg=False):
+  name: str = package['Package']
+  if not name.startswith('microsoft-edge-'):
+    continue
+  channel = name.replace('microsoft-edge-', '')
+  if channel not in latest_packages:
+    latest_packages[channel] = package
+  else:
+    old_package = latest_packages[channel]
+    if old_package.get_version() < package.get_version():  # type: ignore
+      latest_packages[channel] = package
+
+# Generate a Nix expression for each channel
+channel_strs: list[str] = []
+for channel, package in latest_packages.items():
+  match = Version.re_valid_version.match(package['Version'])
+  assert match is not None
+
+  version = match.group('upstream_version')
+  revision = match.group('debian_revision')
+  sri = 'sha256-' + base64.b64encode(bytes.fromhex(package['SHA256'])).decode('ascii')
+
+  channel_strs.append(f'''
+  {channel} = import ./browser.nix {{
+    channel = "{channel}";
+    version = "{version}";
+    revision = "{revision}";
+    sha256 = "{sri}";
+  }};''')
+
+# Write the final expression
+nix_expr = f'''{{{'  '.join(channel_strs)}
+}}'''
+with open('default.nix', 'w') as f:
+  f.write(nix_expr)
```

Comment:
I've replaced the comments by function definitions because comments tend to get outdated. Also, I've added a small confirmation prompt to avoid people accidentally overwriting `default.nix`, fixed some trailing newlines, and added slightly more debug info:
```suggestion
def packages():
  packages_url = 'https://packages.microsoft.com/repos/edge/dists/stable/main/binary-amd64/Packages.gz'
  handle = request.urlopen(packages_url)
  data = gzip.GzipFile(fileobj=handle).read()
  return data

def latest_packages(packages):
  latest_packages: dict[str, Packages] = {}
  for package in Packages.iter_paragraphs(packages, use_apt_pkg=False):
    name: str = package['Package']
    if not name.startswith('microsoft-edge-'):
      continue
    channel = name.replace('microsoft-edge-', '')
    if channel not in latest_packages:
      latest_packages[channel] = package
    else:
      old_package = latest_packages[channel]
      if old_package.get_version() < package.get_version():  # type: ignore
        latest_packages[channel] = package
  return latest_packages

def nix_expressions(latest):
  channel_strs: list[str] = []
  for channel, package in latest.items():
    print(f"Processing {channel} {package['Version']}")
    match = Version.re_valid_version.match(package['Version'])
    assert match is not None

    version = match.group('upstream_version')
    revision = match.group('debian_revision')
    sri = 'sha256-' + base64.b64encode(bytes.fromhex(package['SHA256'])).decode('ascii')

    channel_strs.append(f'''
    {channel} = import ./browser.nix {{
      channel = "{channel}";
      version = "{version}";
      revision = "{revision}";
      sha256 = "{sri}";
    }};''')
  return channel_strs

def write_expression():
  latest = latest_packages(packages())
  channel_strs = nix_expressions(latest)
  nix_expr = f'''{{{''.join(channel_strs)}
  }}'''

  print(f'''
This script will update the default.nix file with the following contents:
{nix_expr}\n
Are you sure you want to continue [y/N]?
  ''')
  response = input('> ')
  if response.lower() != 'y':
    print('Aborting')
    exit(1)
  with open('default.nix', 'w') as f:
    f.write(nix_expr)

write_expression()
```
Output to stdout looks as follows:
```
$ nix-shell update.py

Processing stable 110.0.1587.69-1
Processing beta 111.0.1661.38-1
Processing dev 112.0.1702.3-1

This script will update the default.nix file with the following contents:
{
    stable = import ./browser.nix {
      channel = "stable";
      version = "110.0.1587.69";
      revision = "1";
      sha256 = "sha256-tE5DezWgcxLyF+wdXkppLzYufedD1s5qhG643klPl4Y=";
    };
    beta = import ./browser.nix {
      channel = "beta";
      version = "111.0.1661.38";
      revision = "1";
      sha256 = "sha256-D2HG+O+2jOpPmwVoaaItmqPDdB+NdeQNKdZoEw5mtio=";
    };
    dev = import ./browser.nix {
      channel = "dev";
      version = "112.0.1702.3";
      revision = "1";
      sha256 = "sha256-vfuf7CNgekyLr+58L7UVZIpTcBv48XEDrAUHG1KMcKI=";
    };
  }

Are you sure you want to continue [y/N]?
```

Reply:
Indentation is not perfect yet. Would be great if someone who knows more Python than me could look at that

Reply:
Thank you very much. 

I'm trying to fix indentation following [https://stackoverflow.com/questions/57174866/preserving-indentation-in-a-triple-quoted-fstring](https://stackoverflow.com/questions/57174866/preserving-indentation-in-a-triple-quoted-fstring). I have also formatted the code with autopep8. Now indentation uses 4 spaces as [PEP 8](https://peps.python.org/pep-0008/) suggested.